chanrestore.go

```diff
@@ -48,19 +48,40 @@ type chanDBRestorer struct {
 func (c *chanDBRestorer) openChannelShell(backup chanbackup.Single) (
 	*channeldb.ChannelShell, error) {
 
-	// First, we'll also need to obtain the private key for the shachain
-	// root from the encoded public key.
-	//
-	// TODO(roasbeef): now adds req for hardware signers to impl
-	// shachain...
-	privKey, err := c.secretKeys.DerivePrivKey(backup.ShaChainRootDesc)
-	if err != nil {
-		return nil, fmt.Errorf("unable to derive shachain root key: %v", err)
-	}
-	revRoot, err := chainhash.NewHash(privKey.Serialize())
-	if err != nil {
-		return nil, err
+	// The shachain root that seeds the RevocationProducer for this channel.
+	// It currently has two possible formats.
+	var revRoot *chainhash.Hash
+	var err error
+
+	// If the PubKey field is non-nil, then this shachain root is using the
+	// legacy non-ECDH scheme.
+	if backup.ShaChainRootDesc.PubKey != nil {
+		// Obtain the private key for the shachain root from the encoded public
+		// key.
+		privKey, err := c.secretKeys.DerivePrivKey(backup.ShaChainRootDesc)
+		if err != nil {
+			return nil, err
+		}
+
+		revRoot, err = chainhash.NewHash(privKey.Serialize())
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		// This is the scheme in which the shachain root is derived via an
+		// ECDH operation on the private key of ShaChainRootDesc and the remote
+		// node's public key.
+		ecdh, err := c.secretKeys.ECDH(backup.ShaChainRootDesc, backup.RemoteNodePub)
```

Comment:
Hmm, I had originally envisioned just doing ECDH against one of our _own_ keys. It seems a bit redundant, but this ensure that the remote party isn't able to somehow re-derive _our_ shachain root if they somehow learn of an SCB.

Reply:
It should ideally be one of our keys.

Reply:
Changed to use our `MultiSigKey`